File: src/aoc23/Day15.java
package aoc23;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;

public class Day15 {
    record Lens(String label, int focalLength) {
    }
    public static void main(String[] args) throws IOException {
        String path = "C:/dev/advent-of-code/inputs/t";
        var lines = Files.readString(Paths.get(path)).split(",");
        part1(lines);
        part2(lines);
    }

    private static void part1(String[] lines) {
        System.out.println("Part 1: " + Arrays.stream(lines).mapToInt(Day15::getHash).sum());
    }
    private static void part2(String[] lines) {
        var boxes = new HashMap<Integer, ArrayList<Lens>>(256);
        o:
        for (String line : lines) {
            int op = line.indexOf('-');
            if (op == -1) op = line.indexOf('=');
            String currLabel = line.substring(0, op);
            int h = getHash(currLabel);
            boxes.computeIfAbsent(h, k -> new ArrayList<>());
            if (line.charAt(op) == '-') {
                boxes.get(h).removeIf(l -> l.label.equals(currLabel));
            } else {
                Lens newLens = new Lens(currLabel, line.charAt(op + 1) - '0');
                for (int i = 0; i < boxes.get(h).size(); i++) {
                    if (boxes.get(h).get(i).label.equals(currLabel)) {
                        boxes.get(h).set(i, newLens);
                        continue o;
                    }
                }
                boxes.get(h).add(newLens);
            }
        }
        int ans = 0;
        for (int i: boxes.keySet())
            for (int j = 0; j < boxes.get(i).size(); j++)
                ans += (i + 1) * (j + 1) * boxes.get(i).get(j).focalLength;

        System.out.println("Part 2: " + ans);
    }
    private static int getHash(String l) {
        return l.chars().reduce(0, (acc, c) -> (acc + c) * 17 % 256);
    }
}

File: src/aockt/Day15.kt
package aockt

import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*

object Day15 {
    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val path = "C:/dev/advent-of-code/inputs/t"
        val lines = Files.readString(Paths.get(path)).split(",".toRegex()).toTypedArray()
        part1(lines)
        part2(lines)
    }

    private fun getHash(l: String): Int {
        return l.chars().reduce(0) { acc: Int, c: Int -> (acc + c) * 17 % 256 }
    }

    private fun part1(lines: Array<String>) {
        println("Part 1: " + lines.sumOf { getHash(it) })
    }

    private fun part2(lines: Array<String>) {
        val boxes = Array(256) { mutableListOf<Lens>() }
        for (line in lines) {
            var op = line.indexOf('-')
            if (op == -1) op = line.indexOf('=')
            val label = line.substring(0, op)
            val h = getHash(label)
            if (line[op] == '-') {
                boxes[h].removeIf { it.label == label }
            } else {
                val newLens = Lens(label, line[op + 1].code - '0'.code)
                val idx = boxes[h].indexOfFirst { it.label == label }
                if (idx == -1)
                    boxes[h].add(newLens)
                else
                    boxes[h][idx] = newLens
            }
        }
        var ans = 0
        boxes.forEachIndexed { i, x ->
            x.forEachIndexed { j, y ->
                ans += (i + 1) * (j + 1) * y.focalLength
            }
        }
        println("Part 2: $ans")
    }

    internal data class Lens(val label: String, val focalLength: Int)
}